src/topology/subset_properties.lean

```diff
@@ -627,63 +706,76 @@ end⟩
 /-- The connected component of a point is the maximal connected set
 that contains this point. -/
 def connected_component (x : α) : set α :=
-⋃₀ { s : set α | is_connected s ∧ x ∈ s }
-
-theorem is_connected_connected_component {x : α} : is_connected (connected_component x) :=
-is_connected_sUnion x _ (λ _, and.right) (λ _, and.left)
+⋃₀ { s : set α | is_preconnected s ∧ x ∈ s }
 
 theorem mem_connected_component {x : α} : x ∈ connected_component x :=
-mem_sUnion_of_mem (mem_singleton x) ⟨is_connected_singleton, mem_singleton x⟩
+mem_sUnion_of_mem (mem_singleton x) ⟨is_connected_singleton.is_preconnected, mem_singleton x⟩
+
+theorem is_connected_connected_component {x : α} : is_connected (connected_component x) :=
+⟨⟨x, mem_connected_component⟩, is_preconnected_sUnion x _ (λ _, and.right) (λ _, and.left)⟩
 
-theorem subset_connected_component {x : α} {s : set α} (H1 : is_connected s) (H2 : x ∈ s) :
+theorem subset_connected_component {x : α} {s : set α} (H1 : is_preconnected s) (H2 : x ∈ s) :
   s ⊆ connected_component x :=
 λ z hz, mem_sUnion_of_mem hz ⟨H1, H2⟩
 
 theorem is_closed_connected_component {x : α} :
   is_closed (connected_component x) :=
 closure_eq_iff_is_closed.1 $ subset.antisymm
   (subset_connected_component
-    is_connected_connected_component.closure
+    is_connected_connected_component.closure.is_preconnected
     (subset_closure mem_connected_component))
   subset_closure
 
 theorem irreducible_component_subset_connected_component {x : α} :
   irreducible_component x ⊆ connected_component x :=
 subset_connected_component
-  (is_connected_of_is_irreducible is_irreducible_irreducible_component)
+  is_irreducible_irreducible_component.is_connected.is_preconnected
   mem_irreducible_component
 
-/-- A connected space is one where there is no non-trivial open partition. -/
-class connected_space (α : Type u) [topological_space α] : Prop :=
-(is_connected_univ : is_connected (univ : set α))
+/-- A preconnected space is one where there is no non-trivial open partition. -/
+class preconnected_space (α : Type u) [topological_space α] : Prop :=
+(is_preconnected_univ : is_preconnected (univ : set α))
+
+/-- A connected space is a nonempty one where there is no non-trivial open partition. -/
+class connected_space (α : Type u) [topological_space α] extends preconnected_space α : Prop :=
```

Comment:
Why don't you use a typeclass assumption `[nonempty α]` instead, to get typeclass inference to do some work for you for free when necessary?

Reply:
Because it means that you have to write
```lean
variables (X : Type) [topological_space X] [nonempty X] [connected_space X]
```
and I thought that was overly verbose.

Reply:
OK. Then maybe `extends preconnected_space α, nonempty α`?

Reply:
I tried that, but `nonempty` is not a structure, but an inductive type. So you can't extend it )-;

Reply:
Ah crap. Would changing `nonempty` to a structure with a single field break a lot of things?

Reply:
I didn't have the courage to try that yet.

Reply:
@sgouezel `nonempty` cannot be a structure, because structures must have projections for their fields and `Prop`s don't support that. Of course, the projection for `nonempty` is `choice`, and so making changes to `nonempty` would also officially change the lean axiomatization.

Reply:
You can, however, add `nonempty X` as a field of `connected_space`, and define an instance for it, and it will be about as good as a parent structure.

Reply:
I don't get it. Is there a problem with
```lean
class my_nonempty (α : Type*) : Prop :=
(my_field : ∃(x : α), true)
```

Reply:
No, but that's not the same as the definition of nonempty. Keep in mind that `nonempty` is logically prior to `exists`; Really, `exists x : A, p x` is just `nonempty (Sigma x : A, p x)` but lean defines them both via the general inductive types framework. A direct definition of `nonempty` can't reference `exists`.

Reply:
Sure, that's not the same definition as nonempty, but it's mathematically the same. The question is: what would we lose if we replaced all uses of `nonempty` in the library with `my_nonempty`? (With the gain that a class could now extend `my_nonempty` transparently).

Reply:
Apparently, to do that, you would also need to redefine `exists` and run into issues that touch the axioms of Lean... I.e., we would need a `lean-3.6.0c`.

Reply:
I guess I am not clear. I don't want to change the existing definitions of `exists` (which does not depend on `nonempty`), just replace the definition of `nonempty` with the one above (which is a structure). So, no new axiom or whatever. And I am perfectly ready to hear that this is not a good idea because it will break this or this that depends on `nonempty` being an inductive, if there are such things in the library. And I don't think this is an important issue, by the way: we can definitely go with your strategy of adding a field in the definition of `irreducible` and `connected` that would be defined indepently of `nonempty`.

Reply:
@sgouezel I'm speaking on two media now, see https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/should.20nonempty.20be.20a.20structure/near/187559580 , but If you have `my_nonempty` as a class wrapper around `nonempty`, then yes it's a structure but it will not have the desired behavior. You will get two projections in your structure, one to `my_nonempty` and one to `nonempty`, and you still will not have the instance `nonempty <- foo` that was the point of all this.

Reply:
@jcommelin I do think that `set.nonempty (univ : set α)` is an unnecessarily obsure way to write `nonempty α`, though.

Reply:
Fair enough. I'll fix that.

Reply:
Done.